pac4j-couch/src/main/java/org/pac4j/couch/profile/service/CouchProfileService.java

```diff
@@ -0,0 +1,185 @@
+		CommonHelper.assertNotNull("couchDbConnector", this.couchDbConnector);
+		defaultProfileDefinition(new CommonProfileDefinition<>(x -> new CouchProfile()));
+
+		super.internalInit(context);
+	}
+
+	@Override
+	protected void insert(final Map<String, Object> attributes) {
+		logger.debug("Insert doc: {}", attributes);
+		couchDbConnector.create(attributes);
+	}
+
+	@Override
+	protected void update(final Map<String, Object> attributes) {
+		try {
+			final String id = (String) attributes.get(COUCH_ID);
+			final InputStream oldDocStream = couchDbConnector.getAsStream(id);
+			final JsonNode oldDoc = objectMapper.readTree(oldDocStream);
+			final String rev = oldDoc.get("_rev").asText();
+			attributes.put("_rev", rev);
+			couchDbConnector.update(attributes);
+			logger.debug("Updating id: {} with attributes: {}", id, attributes);
+		} catch (DocumentNotFoundException e) {
+			couchDbConnector.create(attributes);
+		} catch (IOException e) {
+			logger.error("", e);
+		}
+	}
+
+	@Override
+	protected void deleteById(final String id) {
+		logger.debug("Delete id: {}", id);
+		try {
+			final InputStream oldDocStream = couchDbConnector.getAsStream(id);
+			final JsonNode oldDoc = objectMapper.readTree(oldDocStream);
+			final String rev = oldDoc.get("_rev").asText();
+			couchDbConnector.delete(id, rev);
+		} catch (DocumentNotFoundException e) {
+			logger.debug("id {} is not in the database", id);
+		} catch (IOException e) {
+			logger.error("", e);
+		}
+	}
+
+	@Override
+	protected List<Map<String, Object>> read(final List<String> names, final String key, final String value) {
+		logger.debug("Reading key / value: {} / {}", key, value);
+		final List<Map<String, Object>> listAttributes = new ArrayList<>();
+		final TypeReference<HashMap<String,Object>> typeRef = new TypeReference<HashMap<String,Object>>() {};
+		if (key.equals(COUCH_ID)) {
+			try {
+				final InputStream oldDocStream = couchDbConnector.getAsStream(value);
+				final Map<String, Object> res = objectMapper.readValue(oldDocStream, typeRef);
+				final Map<String, Object> newAttributes = new HashMap<>();
+				for (final Map.Entry<String, Object> entry : res.entrySet()) {
+					final String name = entry.getKey();
+					if (names == null || names.contains(name)) {
+						newAttributes.put(name, entry.getValue());
+					}
+				}
+				listAttributes.add(newAttributes);
+			} catch (DocumentNotFoundException e) {
+			} catch (IOException e) {
+				logger.error("", e);
+			}
+		}
+		else {
+			// supposes a by_$key view in the design document, see documentation
+			final ViewQuery query = new ViewQuery()
+					.designDocId("_design/pac4j")
+					.viewName("by_"+key)
+					.key(value);
+			final ViewResult result = couchDbConnector.queryView(query);
+			for (ViewResult.Row row : result.getRows()) {
+				final String stringValue = row.getValue();
+				Map<String, Object> res = null;
+				try {
+					res = objectMapper.readValue(stringValue, typeRef);
```

Comment:
I think this part is the same as at line 118: can we refactor that in some separate `populateAttributes` (or another name if you prefer)  method?